torch/nn/modules/rnn.py

```diff
@@ -200,10 +228,14 @@ def forward(self, input, hx=None):
             # the user believes he/she is passing in.
             hx = self.permute_hidden(hx, sorted_indices)
 
+        is_tranposed = self.use_transposed_weights and not self.training
+        self.transpose_paramters(is_tranposed)
+        _weights = self._transposed_weights if is_tranposed else self.get_flat_weights()
```

Comment:
This is very much non-standard, and we should not hack such a functionality into every module individually without having a consistent library-wide consensus on how to implement support for different weight layouts (this includes convolutional modules in particular).

Reply:
Apart from that, this will *unexpectedly* change the weight format without the user even asking for it, which is definitely not what any good library should do. This should be at least controlled by a flag which is off by default to keep backward compatibility, but we should really figure out a more widely applicable approach.

Reply:
Yes, indeed this is not a decent design, it is mostly a temporary hack.

My complete idea to is isolate mkldnn storage from cpu storage, so we have an `MKLDNNTensorImpl`, kind of like `SparseTensorImpl`, and user may invoke mkldnn by `tensor.to_mkldnn()`, kind of like `tensor.to_sparse()`. This should be no hazard to both autograd and jit.

The mkldnn storage will basically handle two things:
1. mkldnn layout cache between forward and backward during training
2. mkldnn layout cache for inference

It will bring performance boost for `rnn` as well `convolution`, both training and inference.

Any comments on such design? @apaszke @cpuhrsch

Reply:
#16038 documented the discussion on the idea.

Reply:
@mingfeima - this is a fairly complicated topic. Introducing a new tensor type isn't a path well defined for now. So I suggest we add a flag to prevent what @apaszke is pointing out. This is different from enabling/disabling MKL-DNN at runtime however. This could be a flag we pass as an argument and set to false by default. Let's worry about that once the test failures are resolved and we can confirm the performance gains internally.

Reply:
Adding a new tensor type is definitely an option. To be honest I would be quite unhappy even if we were to have a flag that controls this, because it sets an ugly precedent for other cases like this one. We should have a cleaner solution that covers more cases.

Reply:
Totally understand that! Upper level interface should be restricted by library dependent hacking.

You can now consider this PR as experimental to verify the `LSTM` performance on your production models. If the performance is not satisfactory, we will continue working on the mkldnn library part.

We are trying to add a new `TensorTypeId` in `c10` for mkldnn. Once it is done, nothing from `rnn.py` needs to be changed, mkldnn weight caching would be taken care of by the opaque data handle. Thus the mkldnn rnn function would be as clean as cudnn ones.